### PR TITLE
Refactor validate sample sheet

### PR DESCRIPTION
**Pull request by henrikstranneheim**
### Added

- General `validate_file_suffix` function to IO
- Add new `FlowCellMode` constant class

### Changed

- Refactored CLI validate_sample_sheet
- Add new option to specify --flow-cell-mode to `cg demultiplex samplesheet validate`

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b refactor_validate_sample_sheet -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/constants/demultiplexing.py
Comment (henrikstranneheim) [reply cg/constants/demultiplexing.py:97]: I lack the knowledge of which instrument uses each type. I ask the wisdom of @Karl-Svard @karlnyr in this matter!

**Review by diitaz93 — COMMENTED** [review]
Some comments

File: cg/constants/demultiplexing.py
Comment (diitaz93) [reply cg/constants/demultiplexing.py:97]: I talked to @Karl-Svard, and it seems to be more complicated than that,so better discuss it with them

Comment (karlnyr) [reply cg/constants/demultiplexing.py:97]: So I believe that we currently infer the type of sequencer from the parsed RunParameters file which is the "FlowCellMode". So to say that the type is S4, S2 is incorrect - the type should be the sequencer type. With the new sequencing machines, we are swapping flow cells to 5B, 10B, and 25B which most likely will be behind the same tag. To be honest, I think that this information is best fetched from LIMS/Arnold as I would rather rely on the setup steps to determine this rather than the output format of a file that Illumina designs. That would also remove some very "sketchy" xmltree coding which I find un-intuitive. I will however ask Emilia at Illumina about possible changes to any setup files/output files which could affect not only the sample sheet creation but also the parsing of demultiplexing data.

Comment (henrikstranneheim) [reply cg/constants/demultiplexing.py:97]: From cgmodels:
```
sheet_type: Literal["2500", "SP", "S2", "S4"]
novaseqsample = {"bcl2fastq": NovaSeqSampleBcl2Fastq, "dragen": NovaSeqSampleDragen}
sample_type = Sample if sheet_type == "2500" else novaseqsample[bcl_converter]

class NovaSeqSampleBcl2Fastq(NovaSeqSample):
    sample_id: str = Field(..., alias="SampleID")
    project: str = Field(..., alias="Project")

class NovaSeqSample(Sample):
    """This model is used when parsing/validating existing novaseq sample sheets"""

    second_index: str = Field(..., alias="index2")

class SampleSheet(BaseModel):
    type: str
    samples: List[Sample]
```

It seems that the sheet type is coupled to the software on each sequencing instrument and how that instrument expects certain headers to be named. I guess certain flow cell types are able to be processed by more than one instrument, but it should be how the instrument parses the sheet that is the determining factor here.

@karlnyr @Karl-Svard Is it correct to assume that the sheet format per instrument does not change depending on flow cell type?

File: cg/io/validate_psth.py
Comment (henrikstranneheim) [reply cg/io/validate_psth.py:1]: Good catch! Fixed

File: cg/cli/demultiplex/sample_sheet.py
Comment (henrikstranneheim) [reply cg/cli/demultiplex/sample_sheet.py:19]: Fixed

File: cg/cli/demultiplex/sample_sheet.py
Comment (diitaz93) [inline cg/cli/demultiplex/sample_sheet.py:40]: This information can be retrieved from the RunParameters.xlm file in a field called `/RfidsInfo/FlowCellMode`. See for example the [fixture run parameters](https://github.com/Clinical-Genomics/cg/blob/master/tests/fixtures/apps/demultiplexing/flowcell-runs/201203_A00689_0200_AHVKJCDRXX/RunParameters.xml) line 26. I was thinking of including in my next PR a function inside the model RunParameters that returns this information and could be used for this purpose, so we could avoid adding more parameters to the CLI command. What do you think? My (still unfinished) PR is #1995, see file `cg/models/demultiplex/run_parameters.py` line 66. I could make it ready for review so we can merge it and make use of this in here

Comment (henrikstranneheim) [reply cg/cli/demultiplex/sample_sheet.py:40]: Sounds good!I will put this in blocked until #1995 is merged

**Review by diitaz93 — APPROVED** [review]
Looks good 👍 Is this going to be a Major version? as the CLI was modified?

File: cg/models/demultiplex/flow_cell.py
Comment (henrikstranneheim) [reply cg/models/demultiplex/flow_cell.py:143]: Sounds great! Yes - a major it is

**Comment by henrikstranneheim** [comment]
### Validate samplesheet  ✅ 
```
[henrik.stranneheim@hasta:~] [S_base] 2m49s $ cg demultiplex samplesheet validate --help
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Usage: cg demultiplex samplesheet validate [OPTIONS] SHEET

  Validate a sample sheet.

Options:
  --flow-cell-mode [SP|S2|S4|2500]
                                  Instrument sample sheet flow cell mode
                                  [default: S4]
  -b, --bcl-converter [bcl2fastq|dragen]
                                  Specify bcl conversion software. Choose
                                  between bcl2fastq and dragen. Default is
                                  bcl2fastq.
  --help                          Show this message and exit.
[henrik.stranneheim@hasta:~] [S_base] 17s $ cg demultiplex samplesheet validate --flow-cell-mode "S4" -b dragen /home/proj/stage/demultiplexed-runs/221214_A00689_0717_AHWV5LDRX2/Unaligned/SampleSheet.csv
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Validating sample sheet /home/proj/stage/demultiplexed-runs/221214_A00689_0717_AHWV5LDRX2/Unaligned/SampleSheet.csv
Order samples by lane
Validate that samples are unique in lane 1
Validate that samples are unique in lane 2
Sample sheet looks fine
[henrik.stranneheim@hasta:~] [S_base] 18s $ cg demultiplex samplesheet validate -b dragen /home/proj/stage/demultiplexed-runs/221214_A00689_0717_AHWV5LDRX2/Unaligned/SampleSheet.csv
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Validating sample sheet /home/proj/stage/demultiplexed-runs/221214_A00689_0717_AHWV5LDRX2/Unaligned/SampleSheet.csv
Order samples by lane
Validate that samples are unique in lane 1
Validate that samples are unique in lane 2
Sample sheet looks fine
```

### Wrong suffix ✅ 
```
[henrik.stranneheim@hasta:~] [S_base] 18s $ cg demultiplex samplesheet validate -b dragen /home/proj/stage/demultiplexed-runs/221214_A00689_0717_AHWV5LDRX2/Unaligned/Undetermined_S0_L001_R1_001.fastq.gz
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running cg demultiplex.
Validating sample sheet /home/proj/stage/demultiplexed-runs/221214_A00689_0717_AHWV5LDRX2/Unaligned/Undetermined_S0_L001_R1_001.fastq.gz
File in path /home/proj/stage/demultiplexed-runs/221214_A00689_0717_AHWV5LDRX2/Unaligned/Undetermined_S0_L001_R1_001.fastq.gz seems to be in wrong format
Suffix .gz is not .csv
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_main/bin/cg", line 8, in <module>
    sys.exit(base())
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/demultiplex/sample_sheet.py", line 52, in validate_sample_sheet
    validate_file_suffix(path_to_validate=sheet, target_suffix=FileExtensions.CSV)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/io/validate_path.py", line 15, in validate_file_suffix
    raise ValidationError("Incorrect suffix")
cg.exc.ValidationError: Incorrect suffix
```

**Comment by henrikstranneheim** [comment]
```
Log deploy... done.
cg, version 30.1.0
[henrik.stranneheim@hasta:~] [P_base] 3m3s $
```

**Pull request merged**
